Fix handling of overloaded functions and some other issues in SolidityMetadata

stephan-tolksdorf opened the pull request:
This draft PR is my proposed fix for issue #1210. It builds on PR #1211.

This PR currently

1. Rewrites `SolidityMetadata.__init__` to fix the handling of overloaded functions.
2. Separates the abi items for functions, events, constructor and fallback function into separate properties (though the event items are not yet publicly exposed).
3. Removes the `abi` property.
4. Removes the `add_function` method.
5. Adds a decorator for deprecating functions.
6. Deprecates the `functions` and `hashes` properties due to their problematic handling of the fallback function and adds `function_signatures` and `function_selectors` as replacements.
7. Deprecates the `get_hash` method, which seems redundant in its current form.
8. Adds static methods for constructing function and tuple type signatures that can handle ABIv2 tuple arguments.
9. Adds the `constructor_abi`, `has_non_default_constructor` and `has_non_default_fallback_function` properties.
10. Adds some docstrings and type annotations.

I can of course split this PR up or remove changes, change names, etc. if you'd like me to.

If you're OK with the direction of this PR, I could add more tests for `SolidityMetadata`.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/trailofbits/manticore/1212)
<!-- Reviewable:end -->

disconnect3d commented:
@stephan-tolksdorf hey

> This draft PR

If this is a `work in progress`, please add `[WIP]` to the title (and remove it later on + send a message that it is ready for review; also if you have any questions, feel free to ask).

stephan-tolksdorf commented:
@disconnect3d I've updated the title. 

I'd like to add more tests to this PR, but before doing so, it would be great if I could get some feedback on the current version of the PR. In particular I'd like to know whether you're OK with the scope of the PR and how I went about solving the issue.

disconnect3d commented on manticore/ethereum/__init__.py:
```diff
@@ -99,29 +100,28 @@ def will_open_transaction_callback(self, state, tx):
+            for func_hsh in md.function_selectors:
                 abi = md.get_abi(func_hsh)
-                func_name = md.get_func_name(func_hsh)
+                if abi['type'] == 'fallback':
```
:+1: this seems much better than `func_hsh == b'\0\0\0\0'`

disconnect3d commented on manticore/ethereum/__init__.py:
```diff
@@ -99,29 +100,28 @@ def will_open_transaction_callback(self, state, tx):
                 if self._mutability == 'mutable' and abi.get('constant', False):
                     continue
-                if not re.match(self._regexp, func_name):
+                if not re.match(self._regexp, abi['name']):
```
Loud thinking: is using regex here a good idea? It might slow down things a lot (since this is a callback).

@feliam @yan @mossberg what do u think?

disconnect3d commented on manticore/ethereum/abi.py:
```diff
@@ -137,6 +137,8 @@ def _serialize_bytes(value):
     def _serialize_tuple(types, value, dyn_offset=None):
         result = bytearray()
         dyn_result = bytearray()
+        if len(types) != len(value):
```
We should probably use `if __debug__` in such cases as described here https://stackoverflow.com/a/27455446

stephan-tolksdorf replied:
I've implemented the check the same way it's done in `_serialize_array`.

Personally I don't think it's the right choice for a tool like Manticore to trade the safety of an always-on check against the minuscule performance improvement from disabling an assert like this in release mode.

mossberg replied:
imo i think @stephan-tolksdorf 's implementation is fine, and mirroring `_serialize_array` is correct. I actually think asserts and this explicit error checking are semantically different, and not necessarily interchangeable -- this case is more handling bad input, rather than expressing an invariant so I think the explicit check is more semantically correct. i don't think we should care too much about perf in this case. :)

stephan-tolksdorf replied:
@mossberg Good point regarding the difference of checking preconditions on inputs vs checking internal invariants!

disconnect3d commented on manticore/ethereum/solidity.py:
```diff
@@ -1,10 +1,36 @@
+        for c in components:
+            t: str = c['type']
+            if t.startswith('tuple'):
+                assert len(t) == 5 or t[5] == '['
```
Shouldn't it be `len(t) >= 5`? as we use `t[5:]` later on

stephan-tolksdorf replied:
No, since in this branch `len(t)` is always ≥ 5 because `t` starts with 'tuple' and the right side of the `or` expression should only be evaluated if the `t` has more than 5 characters.

disconnect3d replied:
Oh it is `or` here. That makes sense :).

feliam commented:
Hi @stephan-tolksdorf thank you for this. This is a clear improvement towards handling more complex metadata and contracts.  We are going to quick test this PR on our internal scripts list to check that nothing goes super wrong and get back to you. 
BUt other than that I think it is going in a good direction.

stephan-tolksdorf commented:
@feliam That's good to hear, thank you!

Note that this PR contains also the commit from the previous PR #1211. If and when we get to the merging, it would be good if both PRs are merged in the right order.

mossberg commented on manticore/ethereum/abi.py:
```diff
@@ -137,6 +137,8 @@ def _serialize_bytes(value):
         result = bytearray()
         dyn_result = bytearray()
+        if len(types) != len(value):
+            raise ValueError(f"The number of values to serialize is {'less' if len(value) < len(types) else 'greater'} than the number of types")
```
pls use EthereumError for this, we're trying to shift over to that

stephan-tolksdorf replied:
The rest of the code in that file currently uses `ValueError` and I thought that keeping the code locally consistent might be preferable. 

In general, I think it could be quite helpful to use separate exception hierarchies for (Python) programming errors and errors that arise during the normal EVM execution (runtime errors due to e.g. IO errors would be third category). Since `EthereumError` is a subclass of `ManticoreError`, using it for all exceptions in the Ethereum code base increases the risk that programming errors are inadvertently masked. 

For example, currently the Executor logs and suppresses all exceptions, even asserts, which can make it more difficult to diagnose errors and test fixes. Using separate exception hierarchies could make it easier to let exceptions caused by programming errors bubble up to the parent Manticore process/instance.

mossberg commented:
i debugged the test failures; some of the tests were passing a symbolic argument when the constructor didn't take one. we didn't previously error, but now we do, so the tests failed. this patch fixes it:

```
diff --git a/tests/eth_benchmark.py b/tests/eth_benchmark.py
index a953419..5291621 100644
--- a/tests/eth_benchmark.py
+++ b/tests/eth_benchmark.py
@@ -41,7 +41,7 @@ class EthBenchmark(unittest.TestCase):
         self.mevm=None
         shutil.rmtree(self.worksp)

-    def _test(self, name, should_find):
+    def _test(self, name, should_find, use_ctor_sym_arg=False):
         """
         Tests DetectInvalid over the consensys benchmark suite
         """
@@ -52,8 +52,12 @@ class EthBenchmark(unittest.TestCase):

         filename = os.path.join(THIS_DIR, 'binaries', 'benchmark', f'{name}.sol')

+        if use_ctor_sym_arg:
+            ctor_arg = (mevm.make_symbolic_value(),)
+        else:
+            ctor_arg = ()

-        mevm.multi_tx_analysis(filename, contract_name='Benchmark', args=(mevm.make_symbolic_value(),))
+        mevm.multi_tx_analysis(filename, contract_name='Benchmark', args=ctor_arg)

         expected_findings = set(( (c, d) for b, c, d in should_find))
         actual_findings = set(( (c, d) for a, b, c, d in mevm.global_findings))
@@ -66,10 +70,10 @@ class EthBenchmark(unittest.TestCase):
         self._test('assert_constructor', {(23, 'INVALID instruction', True)})

     def test_assert_multitx_1(self):
-        self._test('assert_multitx_1', set())
+        self._test('assert_multitx_1', set(), True)

     def test_assert_multitx_2(self):
-        self._test('assert_multitx_2', {(150, 'INVALID instruction', False)})
+        self._test('assert_multitx_2', {(150, 'INVALID instruction', False)}, True)

     def test_integer_overflow_minimal(self):
         self._test('integer_overflow_minimal', {(163, 'Unsigned integer overflow at SUB instruction', False)})

```

mossberg commented:
i went ahead and applied the diff to the PR branch.

stephan-tolksdorf commented:
@mossberg Thank you!

Regarding deprecations: You could have a policy that all deprecated methods are removed in the next release, or whenever they become a maintenance burden. I like the deprecation notices because they both serve as documentation for the API changes and make it easier for users to upgrade their code bases, especially in cases like these where the property names are quite common and there are subtle differences in behaviour between the old and the new methods. If you still want them to be removed, I can do that of course.

mossberg commented:
> We are going to quick test this PR on our internal scripts to check that nothing goes super wrong and get back to you.

Internal scripts ran just fine with this 👍

stephan-tolksdorf commented:
In the commits I just pushed I've fixed the `get_func_argument_types` implementations, added type checks for the function selector arguments and added more tests.

The type checks use `ValueError`, like in `parse_tx`. If @mossberg confirms that he prefers `EthereumError` everywhere, I'll replace `ValueError` with `EthereumError`.

stephan-tolksdorf commented:
I'll take a look at the failing `eth_detectors.py` tests.

mossberg commented:
@stephan-tolksdorf thanks for the clear explanations :)

re: Errors: I retract my opinion on EthereumError vs builtin errors. I've thought about it more, and now think that since the errors are very "python-y" and _not domain specific_, it does make sense for us to raise a builtin error here vs an error from our custom hierarchy. Furthermore, these are not errors that should ever be caught/handled really...they indicate improper API usage that should be corrected for the right types. For these reasons I think builtin errors are the way to go here. I will say that for all the error checks that are type/isinstance related, `TypeError` is probably a better one to use than `ValueError`. thanks for making me thing more deeply on this!

re: Deprecation: I think you also make great points. It's not as if the maintenance burden of carrying deprecated functions is high, and it's probably worth it for the end user experience. let's leave them in.

> I'll take a look at the failing `eth_detectors.py` tests.

From my brief look, these failures might be the exact same issues that my previous patch above fixed , just in a different set of tests. (various instances in the tests where we provide ctor arguments to contracts that do not take any). I should have actually run the full eth tests locally 😭 it's 1am in my time zone so i can't look at it now, but they should be straightforward to fix. once tests are passing, lgtm to merge.

stephan-tolksdorf commented:
I agree that `TypeError` is more appropriate than `ValueError` for the type checks. Thanks for spotting that! I've replaced all `ValueError` uses in `solidity.py`. 

You were also right about the `eth_detectors.py` tests, which I've fixed the same way you fixed the `eth_benchmark.py` tests.
